rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,97 @@
+use crate::logger::{self, Level};
+use std::fs::{self, File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process;
+
+fn remove_lock(lock_path: &Path) {
+    fs::remove_file(lock_path).ok();
+    log!(
+        Level::Debug,
+        "FsLock: removed lockfile {}",
+        lock_path.display()
+    );
+}
+
+#[derive(Default)]
+pub struct FsLock {
+    lock_path: PathBuf,
+    lock_file: Option<File>,
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.lock_file.is_some() {
+            remove_lock(&self.lock_path);
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock(&mut self, new_lock_path: &Path, pid: &mut libc::pid_t) -> bool {
+        if self.lock_file.is_some() && self.lock_path == new_lock_path {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_path.display()
+        );
+
+        // first we open (and possibly create) the lock file
+        let mut openoptions = OpenOptions::new();
+        openoptions.read(true).write(true).create(true);
+        let mut file = match openoptions.open(&new_lock_path) {
+            Ok(file) => file,
+            Err(_) => return false,
+        };
+
+        // then we lock it (returns immediately if locking is not possible)
+        if unsafe { libc::lockf(file.as_raw_fd(), libc::F_TLOCK, 0) } == 0 {
+            log!(
+                Level::Debug,
+                "FsLock: locked `{}', writing PID...",
+                new_lock_path.display()
+            );
+            let pid = process::id().to_string();
+            let buf = pid.as_bytes();
+            let success = file.write_all(&buf).is_ok();
```

Comment:
The original code also called `ftruncate()`, but this doesn't. I believe this might lead to wrong data in the lockfile in scenarios like this:

1. some Newsboat instance writes a long PID into the file, e.g. "987654321";
2. that Newsboat instance crashes, so the file persists but isn't locked anymore;
3. a new Newsboat instance with a short PID starts, e.g. "123";
4. it locks the file and writes its PID into it;
5. another Newsboat instance starts, fails to lock the file, and displays an error message with a mix of the old and new PIDs.

Please truncate the file before writing into it.

Reply:
Ah, I thought it automatically truncates on write?

Reply:
I don't expect it does, otherwise the file would only contain what the last write put there.